nitransforms/nonlinear.py

```diff
@@ -90,70 +118,162 @@ def from_filename(cls, filename, fmt="X5"):
-#             first = int(np.floor(knots_ijk[dim] + offset) - self._order // 2)
-#             neighbors.append(list(range(first, first + self._order + 1)))
-
-#         # Get indexes of the neighborings clique
-#         ndindex = np.moveaxis(
-#             np.array(np.meshgrid(*neighbors, indexing='ij')), 0, -1).reshape(
-#             -1, self.ndim)
-
-#         # Calculate the tensor B-spline weights of each neighbor
-#         # weights = np.prod(vbspl(ndindex - knots_ijk), axis=-1)
-#         ndindex = [tuple(v) for v in ndindex]
-
-#         # Retrieve coefficients and deal with boundary conditions
-#         zero = np.zeros(self.ndim)
-#         shape = np.array(self._knots.shape)
-#         coeffs = []
-#         for ijk in ndindex:
-#             offbounds = (zero > ijk) | (shape <= ijk)
-#             coeffs.append(
-#                 self._coeffs[ijk] if not np.any(offbounds)
-#                 else [0.0] * self.ndim)
-
-#         # coords[:3] += weights.dot(np.array(coeffs, dtype=float))
-#         return self.reference.inverse.dot(np.hstack((coords, 1)))[:3]
-
-#     def _map_voxel(self, index, moving=None):
-#         """Apply ijk' = f_ijk((i, j, k)), equivalent to the above with indexes."""
-#         return tuple(self._moving[index + self.__s])
+
+class BSplineFieldTransform(TransformBase):
+    """Represent a nonlinear transform parameterized by BSpline basis."""
+
+    __slots__ = ['_coeffs', '_knots', '_weights', '_order', '_moving']
+
+    def __init__(self, coefficients, reference=None, order=3):
+        """Create a smooth deformation field using B-Spline basis."""
+        super(BSplineFieldTransform, self).__init__()
+        self._order = order
+
+        coefficients = _ensure_image(coefficients)
+
+        self._coeffs = np.asanyarray(coefficients.dataobj)
+        self._knots = ImageGrid(four_to_three(coefficients)[0])
+        self._weights = None
+        if reference is not None:
+            self.reference = reference
+
+            if coefficients.shape[-1] != self.ndim:
+                raise ValueError(
+                    'Number of components of the coefficients does '
+                    'not match the number of dimensions')
+
+    def to_field(self, reference=None, dtype="float32"):
+        """Generate a displacements deformation field from this B-Spline field."""
+        reference = _ensure_image(reference)
+        _ref = self.reference if reference is None else SpatialReference.factory(reference)
+        if _ref is None:
+            raise ValueError("A reference must be defined")
+
+        ndim = self._coeffs.shape[-1]
+
+        # If locations to be interpolated are on a grid, use faster tensor-bspline calculation
+        if self._weights is None:
+            self._weights = grid_bspline_weights(_ref, self._knots)
+
+        field = np.zeros((_ref.npoints, ndim))
+
+        for d in range(ndim):
+            field[:, d] = self._coeffs[..., d].reshape(-1) @ self._weights
```

Comment:
Mostly a note to myself... it's hard to keep track of these

```suggestion
            #  1 x Nvox :                          (1 x K) @ (K x Nvox)
            field[:, d] = self._coeffs[..., d].reshape(-1) @ self._weights
```